File: BOJ/BOJ/previous_by_rank/Gold2/boj 2481 해밍 경로.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <unordered_map>
#include <bitset>
#include <string>

using namespace std;

int path[100001];
vector<int> graph_by_id[100001];
unordered_map<int, int> num_to_id;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int N, K;
    cin >> N >> K;
    

    for (int i = 1; i <= N; ++i) {
        string binary;
        cin >> binary;
        int n = stoi(binary, nullptr, 2);
        num_to_id[n] = i;
    }

    for(auto it = num_to_id.begin(); it != num_to_id.end(); ++it) {
        int n, id;
        n = it->first;
        id = it->second;
        for (int j = 0; j < K; ++j) {
            int bit = 1 << j;
            int nxt = n ^ bit;
            if (num_to_id.find(nxt) != num_to_id.end()) {
                int nxt_id = num_to_id[nxt];
                graph_by_id[id].push_back(nxt_id);
            }
        }
    }

    queue<int> q;
    q.push(1);

    while (!q.empty()) {
        int id = q.front();
        q.pop();
        for (int nxt_id : graph_by_id[id]) {
            if (path[nxt_id] == 0) {
                path[nxt_id] = id;
                q.push(nxt_id);
            }
        }
    }

    int M;
    cin >> M;
    for (int i = 0; i < M; ++i) {
        int id;
        cin >> id;
        if (path[id] == 0) {
            cout << -1 << '\n';
            continue;
        }
        vector<int> ans;
        while (id != 1) {
            ans.push_back(id);
            id = path[id];
        }
        ans.push_back(1);
        for (auto it = ans.rbegin(); it != ans.rend(); ++it) {
            cout << *it << ' ';
        }
        cout << '\n';
    }

    return 0;
}

File: BOJ/BOJ/previous_by_rank/Gold2/boj 2481 해밍 경로.py
import sys
from collections import deque

input = sys.stdin.readline
N, K = map(int, input().split())

nums = [-1 for _ in range(N+1)] # 파이썬 dict은 순서가 있던가? 
num_to_id = dict() #
graph_by_id = [list() for _ in range(N+1)]

for i in range(1,N+1):
    n = int(input(),2)
    num_to_id[n] = i
    nums[i] = n

for i in range(1,N+1):
    n = nums[i]
    for j in range(K):
        bit = 1 << j
        nxt = n ^ bit
        if nxt in num_to_id:
            nxt_id = num_to_id[nxt]
            graph_by_id[i].append(nxt_id)

q = deque()
q.append(1)
path = [-1 for _ in range(N+1)]
while q:
    id = q.popleft()
    for nxt_id in graph_by_id[id]:
        if path[nxt_id] == -1: # visited == False 임.
            path[nxt_id] = id # 이전값 저장, 거꾸로 추적하기 위함.
            q.append(nxt_id)

M = int(input())
for _ in range(M):
    id = int(input())
    if path[id] == -1:
        print(-1)
        continue
    ans = []
    while id != 1:
        ans.append(id)
        id = path[id]
    ans.append(1)
    print(*ans[::-1])